src/rhash/hash.c

```diff
@@ -1205,59 +1251,113 @@ static int rc_hash_wii(md5_state_t* md5,
-    if (quad_buffer[0] == 0x5D && quad_buffer[1] == 0x1C && quad_buffer[2] == 0x9E && quad_buffer[3] == 0xA3)
+    rc_file_close(iterator, file_handle);
+    return rc_hash_iterator_error(iterator, "Not a supported WiiWare wad");
+  }
+
+  rc_file_seek(iterator, file_handle, 0x08, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  cert_chain_size =
+    ((uint32_t)quad_buffer[0] << 24) | ((uint32_t)quad_buffer[1] << 16) | ((uint32_t)quad_buffer[2] << 8) | (uint32_t)quad_buffer[3];
+  /* Round up to nearest 0x40 */
+  cert_chain_size = (cert_chain_size + 0x3F) & ~0x3F;
+  rc_file_seek(iterator, file_handle, 0x10, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  ticket_size =
+    ((uint32_t)quad_buffer[0] << 24) | ((uint32_t)quad_buffer[1] << 16) | ((uint32_t)quad_buffer[2] << 8) | (uint32_t)quad_buffer[3];
+  ticket_size = (ticket_size + 0x3F) & ~0x3F;
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  tmd_size =
+    ((uint32_t)quad_buffer[0] << 24) | ((uint32_t)quad_buffer[1] << 16) | ((uint32_t)quad_buffer[2] << 8) | (uint32_t)quad_buffer[3];
+  tmd_size = (tmd_size + 0x3F) & ~0x3F;
+
+  tmd_start_addr = 0x40 + cert_chain_size + ticket_size;
+  if (tmd_size > MAX_BUFFER_SIZE)
+    tmd_size = MAX_BUFFER_SIZE;
+  buffer = (uint8_t*)malloc(tmd_size);
+
+  /* Hash TMD */
+  rc_file_seek(iterator, file_handle, tmd_start_addr, SEEK_SET);
+  rc_file_read(iterator, file_handle, buffer, tmd_size);
+  rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte TMD", tmd_size);
+  md5_append(&md5, buffer, tmd_size);
+  free(buffer);
+
+  /* Get count of content sections */
+  rc_file_seek(iterator, file_handle, (uint64_t)tmd_start_addr + 0x1de, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 2);
+  content_count = ((uint32_t)quad_buffer[0] << 8) | (uint32_t)quad_buffer[1];
+  rc_hash_iterator_verbose_formatted(iterator, "Hashing %u content sections", content_count);
+  content_addr = tmd_start_addr + tmd_size;
+  for (ix = 0; ix < content_count; ix++)
+  {
+    /* Get content section size */
+    rc_file_seek(iterator, file_handle, (uint64_t)tmd_start_addr + 0x1e4 + 8 + ix * 0x20, SEEK_SET);
+    rc_file_read(iterator, file_handle, oct_buffer, 8);
+    content_size = 0;
+    for (jx = 0; jx < 8; jx++)
     {
-      success = rc_hash_wii(&md5, iterator, file_handle);
-    } else {
-      success = rc_hash_iterator_error(iterator, "Not a supported Nintendo disc");
+      content_size <<= 8;
+      content_size += oct_buffer[jx];
     }
+    content_size = (content_size + 0x0F) & ~0x0F;
+    if (content_size > MAX_BUFFER_SIZE)
+      content_size = MAX_BUFFER_SIZE;
+    /* Hash content */
+    buffer = (uint8_t*)malloc(content_size);
+    rc_file_seek(iterator, file_handle, content_addr, SEEK_SET);
+    rc_file_read(iterator, file_handle, buffer, content_size);
+    md5_append(&md5, buffer, content_size);
+    content_addr += content_size;
+    content_addr = (content_addr + 0x3F) & ~0x3F;
```

Comment:
Minor detail: You're applying a ceiling to the address, not the content size. If the address was not already aligned to an 0x40 boundary, this will introduce an unintended drift.

Reply:
All addresses in WAD files are aligned to 0x40. If `content_addr` were to somehow become not aligned to 0x40 (which shouldn't be possible the way the code is written), then keeping it misaligned seems like the worse option to me.